### PR TITLE
Create parallel data structure for types to fasten up lookups later.

### PR DESCRIPTION
**Pull request by sqeezy**
When profiling application build times I noticed that a lot of my performance went into a List.Contains call.
This seems like a problem that's easily solved. I'm well aware that I might have missed some detail here, but this kind of change would bring a lot of performance to applications registering a lot of components.

This is just a rough draft, but I would be very much interested in your input and explanations.

Thank you!

### CODE REVIEW

File: src/Castle.Windsor/Core/ComponentModel.cs [hunk 1]
```diff
@@ -331,6 +332,12 @@ public IEnumerable<Type> Services
 		{
 			get { return services; }
 		}
+		
+		[DebuggerDisplay("Count = {services.Count}")]
+		public HashSet<Type> FastServices
+		{
+			get { return servicesFast; }
+		}
```
Comment (jonorossi) [inline src/Castle.Windsor/Core/ComponentModel.cs:340]: This is on the public interface, I don't think that is right. What sort of performance improvement are you seeing?

Could ComponentModel just have a ContainsService (or similar name) method so the internals aren't exposed?

**Review by jonorossi — CHANGES_REQUESTED** [review]

File: src/Castle.Windsor/Core/ComponentModel.cs (hunk 1 above)
Comment (sqeezy) [reply src/Castle.Windsor/Core/ComponentModel.cs]: I agree that this should not be part of the public api. I just removed it from there. I just internalized the collection for now. Exposing the collection itself internally seems necessary. For that look into [ComponentServicesUtil](https://github.com/sqeezy/Windsor/blob/faster_type_lookup/src/Castle.Windsor/Core/Internal/ComponentServicesUtil.cs#L25).

The root of my problems is in [AbstractHandler](https://github.com/sqeezy/Windsor/blob/faster_type_lookup/src/Castle.Windsor/MicroKernel/Handlers/AbstractHandler.cs#L186). When registering a lot of components the main effort of our application build goes into the Contains implementation of List<T>, which does not scale well. That effort is reduced to almost none with a parallel HashSet<T> in use.

File: src/Castle.Windsor/Core/ComponentModel.cs [hunk 2]
```diff
@@ -40,6 +40,7 @@ public sealed class ComponentModel : GraphNode
 
 		[DebuggerBrowsable(DebuggerBrowsableState.Never)]
 		private readonly List<Type> services = new List<Type>(4);
+		private HashSet<Type> servicesFast = new HashSet<Type>();
```
Comment (jonorossi) [inline src/Castle.Windsor/Core/ComponentModel.cs:43]:
```suggestion
		private readonly HashSet<Type> servicesLookup = new HashSet<Type>();
```

**Review by jonorossi — CHANGES_REQUESTED** [review]
Makes sense with the util method being used from 2 places.

Naming wise, let's go with the purpose rather than "fast" to keep it consistent.

Watch out for code formatting to kept things consistent.

**Comment by sqeezy** [comment]
All your proposed changes are commited.
I think the formatting did not get changed in the process.

**Review by jonorossi — APPROVED** [review]

**Pull request merged**